src/routes/auth.rs

```diff
@@ -75,31 +81,74 @@ impl<'a, 'r> FromRequest<'a, 'r> for User {
             None => return Error::from(ErrorKind::Unknown).into_outcome()
         };
 
-        // Check if a key is required for authentication
-        if !auth_data.key_required() {
-            return Outcome::Success(User::create_and_store_user(request, &auth_data));
-        }
+        let key_opt = request.headers().get_one(AUTH_HEADER);
+        let ldap_config = auth_data.ldap.clone();
+        if ldap_config.enabled {
+            println!("LDAP is enabled");
+            match key_opt {
+                Some(key) => {
+                    match request.headers().get_one(AUTH_HEADER_USERNAME) {
+                        Some(username) => {
+                            let ldap_address = ldap_config.address;
+                            println!("LDAP address: {}, user: {}", ldap_address, username);
+                            let conn = LdapConn::new(&ldap_address);
+                            match conn {
+                                Ok(conn) => {
+                                    let bind_dn = ldap_config.bind_dn.replace("{}", username);
+                                    let result = conn.simple_bind(&bind_dn, key);
+                                    match result {
+                                        Ok(bind_result) => {
+                                            match bind_result.success() {
+                                                Ok(_) => {
+                                                    return Outcome::Success(User::create_and_store_user(request, &auth_data))
+                                                },
+                                                Err(e) => {
+                                                    println!("LDAP bind result: {}", e);
+                                                }
+                                            };
+                                        },
+                                        Err(e) => {
+                                             println!("LDAP bind error: {}", e);
+                                        }
+                                    };
+                                    let _ = conn.unbind();
+                                },
+                                Err(e) => {
+                                    println!("LDAP connection error: {}", e);
+                                }
+                            };
+                        },
+                        None => ()
+                    };
+                 },
+                 None => ()
+            };
```

Comment:
This is a pyramid of code (very indented). Try using features like `map_err` and `?` to reduce the huge amounts of indentation and branching. This will also make the code easier to read and follow.

It would also be a good idea to move the LDAP mechanism into its own method and call it here.

Reply:
This code should be moved to a method which returns `Result<User, Error>`. Then it would be easier to expose errors:
```rust
fn ldap_login(...) -> Result<User, Error> {
    // ...
    conn.simple_bind(&bind_dn, key)
        .context(ErrorKind::Ldap)?
        .success()
        .context(ErrorKind::Ldap)?;
    Ok(User::create_and_store_user(request, &auth_data))
}

match ldap_login(...) {
    Ok(user) => Outcome::success(user),
    Err(e) => e.into_outcome()
}
```

One or more `ErrorKind` variants should also be made to track LDAP errors.